pallets/gear/src/runtime_api.rs

```diff
@@ -40,81 +42,134 @@ impl<T: Config> Pallet<T>
-    ) -> Result<GasInfo, Vec<u8>> {
+        allowance_multiplier: u64,
+    ) -> Result<ReadOnlyReply, String> {
+        // Enabling lazy-pages for this thread.
         Self::enable_lazy_pages();
 
+        // Clearing queue.
+        QueueOf::<T>::clear();
+
+        // Calculating gas allowance for a whole operation,
+        // according to allowance multiplier.
+        let gas_allowance = allowance_multiplier.saturating_mul(BlockGasLimitOf::<T>::get());
+
+        // Updating gas allowance with calculated value.
+        Self::update_gas_allowance(gas_allowance);
+
+        // Casting types into runtime assoc-s.
         let origin = origin.cast();
         let value = value.unique_saturated_into();
 
-        let origin_balance = CurrencyOf::<T>::free_balance(&origin);
+        // Preparing origin balance for extrinsic expenses.
+        let who = Self::prepare_origin_account(origin, gas_limit, value);
 
-        let value_for_gas =
-            <T as pallet_gear_bank::Config>::GasMultiplier::get().gas_to_value(initial_gas);
+        // Executing `send_message` call.
+        Self::send_message(who.into(), destination, payload, gas_limit, value, false)
+            .map_err(|e| Self::dispatch_err_to_string("send_message", e))?;
 
-        let required_balance = CurrencyOf::<T>::minimum_balance()
-            .saturating_add(value_for_gas)
-            .saturating_add(value);
+        // Looking up queue head for message id sent above.
+        let (message_id, _) = Self::queue_head()?;
 
-        let _ = CurrencyOf::<T>::deposit_creating(
-            &origin,
-            required_balance.saturating_sub(origin_balance),
-        );
+        // Creating new manager for queue processing.
+        let mut ext_manager = ExtManager::<T>::default();
 
-        let who = frame_support::dispatch::RawOrigin::Signed(origin);
+        // Creating builtin dispatcher for queue processing.
+        let (builtin_dispatcher, _) = T::BuiltinDispatcherFactory::create();
 
-        QueueOf::<T>::clear();
+        // Queue processing loop.
+        //
+        // Running queue head message if exists.
+        while let Some((_, journal, _)) =
+            Self::dequeue_head_and_run(&mut ext_manager, &builtin_dispatcher, None)?
+        {
+            // Looking through all notes in order to find required reply.
+            for note in &journal {
+                if let JournalNote::SendDispatch { dispatch, .. } = note {
+                    if let Some((replied_to, code)) =
+                        dispatch.reply_details().map(|d| d.into_parts())
+                    {
+                        if replied_to == message_id {
+                            return Ok(ReadOnlyReply {
+                                payload: dispatch.payload_bytes().to_vec(),
+                                value: dispatch.value(),
+                                code,
+                            });
+                        }
+                    }
+                }
```

Comment:
insubstantial but for readability
```suggestion
                let JournalNote::SendDispatch { dispatch, .. } = note else {
                    continue;
                };
                
                let Some((replied_to, code)) =
                    dispatch.reply_details().map(|d| d.into_parts()) else
                {
                    continue;
                };
                
                if replied_to == message_id {
                    return Ok(ReadOnlyReply {
                        payload: dispatch.payload_bytes().to_vec(),
                        value: dispatch.value(),
                        code,
                    });
                }
```